internal/fs/implicit_dirs_test.go

```diff
@@ -569,3 +573,191 @@ func (t *ImplicitDirsTest) AtimeCtimeAndMtime() {
 	ExpectThat(ctime, timeutil.TimeNear(mountTime, delta))
 	ExpectThat(mtime, timeutil.TimeNear(mountTime, delta))
 }
+
+func verifyFileInfo(fi fs.FileInfo, name string, isDir bool) {
+	AssertNe(nil, fi)
+	AssertEq(name, fi.Name())
+	AssertEq(isDir, fi.IsDir())
+}
+
+// Create objects in implicit directories with
+// unsupported object names and
+// test the behavior of stat and ReadDirPicky on the different directories.
+func (t *ImplicitDirsTest) UnsupportedDirNames() {
+	var fi os.FileInfo
+	var entries []os.FileInfo
+	var err error
+
+	// Set up contents.
+	AssertEq(
+		nil,
+		t.createObjects(
+			map[string]string{
+				"foo//bar": "", // unsupported
+				"foo/1":    "", // supported
+				"a/2":      "", // supported
+				"a//2/6":   "", //unsupported
+				"a//3":     "", // unsupported
+				"4":        "", // supported
+				"/4/7":     "", //unsupported
+				"/bar":     "", // unsupported
+				"bar//5":   "", // unsupported
+				"/":        "", //unsupported
+			}))
+
+	// Statting the mount directory should return a directory entry.
+	fi, err = os.Stat(mntDir)
+	AssertEq(nil, err)
+	AssertNe(nil, fi)
+	ExpectTrue(fi.IsDir())
+
+	// Statting the mount-directory/foo should return a directory entry named "foo".
+	fi, err = os.Stat(path.Join(mntDir, "foo"))
+	AssertEq(nil, err)
+	verifyFileInfo(fi, "foo", true)
+
+	// Statting the mount-directory/foo/1 should return a file entry named "1".
+	fi, err = os.Stat(path.Join(mntDir, "foo/1"))
+	AssertEq(nil, err)
+	verifyFileInfo(fi, "1", false)
+
+	// Statting the mount-directory/a should return a directory entry named "a".
+	fi, err = os.Stat(path.Join(mntDir, "a"))
+	AssertEq(nil, err)
+	verifyFileInfo(fi, "a", true)
+
+	// Statting the mount-directory/a/2 should be recognized as a file.
+	fi, err = os.Stat(path.Join(mntDir, "a/2"))
+	AssertEq(nil, err)
+	verifyFileInfo(fi, "2", false)
+
+	// Statting the mount-directory/a//3 should fail as it should be ignored.
+	_, err = os.Stat(path.Join(mntDir, "a/3"))
+	AssertNe(nil, err)
+
+	// Statting the mount-directory/4 should return a file entry named "4".
+	fi, err = os.Stat(path.Join(mntDir, "4"))
+	AssertEq(nil, err)
+	verifyFileInfo(fi, "4", false)
+
+	// Statting the mount-directory/bar should return a directory.
+	fi, err = os.Stat(path.Join(mntDir, "bar"))
+	AssertEq(nil, err)
+	verifyFileInfo(fi, "bar", true)
+
+	// ReadDirPicky on mountdir should not fail as the unsupported sub-directories should be ignored.
+	entries, err = fusetesting.ReadDirPicky(mntDir)
+	AssertEq(nil, err)
+	AssertNe(nil, entries)
+	AssertEq(4, len(entries))
+	verifyFileInfo(entries[0], "4", false)
+	verifyFileInfo(entries[1], "a", true)
+	verifyFileInfo(entries[2], "bar", true)
```

Comment:
> Is this expected? There is no valid object having "bar" directory.

Yes. This is the cost of using MaxResults=1 for lookUpInode. 

If we have objects `bar//1`, and `bar/2` in a bucket, gcs list call during lookUpInode with MaxResults=1 will return `bar//1` . If we reject unsupported object `bar//1` at the time of lookup-inode, then `bar` will never be identified as a directory by GCSFuse, and thus `bar/2` will never be accessible through GCSFuse, which is no better than the current bug. Now, even if there is no object named `bar/2`, we can not know from gcs list call with MaxResults=1, and we end up having to identify `bar` as a directory, albeit empty.

Reply:
> I think this will cause issues in case of RemoveDir operation as the object won't be listed during the operation

I'm not sure I understand what you mean by issues. Are you saying that `bar//1` will not be deleted, or that it'll be deleted without being listed ?

I'll test out RemoveDir manually to see what the behaviour is.

Reply:
> I'll test out RemoveDir manually to see what the behaviour is.

I tested this out manually, in the above example, if I try and delete/rename `bar` in the gcsfuse mount, it only deletes/renames `bar/2` and has no impact on `bar//1` . This looks like the correct behaviour to me as we intend to have `bar//1` be invisible to a GCSFuse mount.